File: built-in plugins dsp/mb-filter-bp.cpp
/**
 * MB Band-Pass Filter
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Band-pass filter for isolating frequency ranges
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FILTER_BP_H
#define MB_FILTER_BP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFilterBp : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-filter-bp";
    static constexpr const char* PLUGIN_NAME    = "MB Band-Pass Filter";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float center = 1000f;  // range [20, 20000]
    float bandwidth = 1f;  // range [0.1, 10]
    float gain = 0f;  // range [-12, 12]
    float mix = 1f;  // range [0, 1]
    };

    MbFilterBp() = default;
    ~MbFilterBp() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.center = std::clamp(params.center, 20f, 20000f);
        params.bandwidth = std::clamp(params.bandwidth, 0.1f, 10f);
        params.gain = std::clamp(params.gain, -12f, 12f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Band-Pass Filter
        return input;
    }
};

#endif // MB_FILTER_BP_H

File: built-in plugins dsp/mb-filter-bp.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbFilterBpPlugin: PluginDefinition = { id: 'mb-filter-bp', slug: 'mb-filter-bp', name: 'MB Band-Pass Filter', category: 'effect', type: 'eq' as any, version: '1.0.0', description: 'Band-pass filter for isolating frequency ranges', author: 'Max Booster', grade: 'A', parameters: [{ id: 'center', name: 'Center Freq', type: 'float', defaultValue: 1000, minValue: 20, maxValue: 20000, automatable: true }, { id: 'bandwidth', name: 'Bandwidth', type: 'float', defaultValue: 1, minValue: 0.1, maxValue: 10, automatable: true }, { id: 'gain', name: 'Gain', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { center: 1000, bandwidth: 1, gain: 0, mix: 1 } };

export default MbFilterBpPlugin;
